ath79: convert drivers into modules

olemmela opened the pull request:
Changing device related drivers into modules allows devices to be gradually converted to DSA without breaking other devices.

The size of the kernel image is also reduced by 94 KB when the drivers are compiled into modules.

olemmela committed:
kernel: netdevices: add dsa support

Add support for building DSA switch drivers as kernel modules.

Signed-off-by: Oskari Lemmela <oskari@lemmela.net>
Link: https://github.com/openwrt/openwrt/pull/15926
Signed-off-by: Hauke Mehrtens <hauke@hauke-m.de>

olemmela committed:
ath79: rename downstream ag71xx driver

Rename the downstream ag71xx driver to ag71xx-legacy.
It allows both upstream and downstream drivers to be compiled into modules.

Signed-off-by: Oskari Lemmela <oskari@lemmela.net>
Link: https://github.com/openwrt/openwrt/pull/15926
Signed-off-by: Hauke Mehrtens <hauke@hauke-m.de>

olemmela committed:
kernel: netdevices: add ag71xx modules

Add support for building both upstream and downstream ag71xx drivers
as modules.

Signed-off-by: Oskari Lemmela <oskari@lemmela.net>
Link: https://github.com/openwrt/openwrt/pull/15926
Signed-off-by: Hauke Mehrtens <hauke@hauke-m.de>

olemmela committed:
kernel: netdevices: add phy modules

Add support for building multiple phy drivers as kernel modules

Signed-off-by: Oskari Lemmela <oskari@lemmela.net>
Link: https://github.com/openwrt/openwrt/pull/15926
Signed-off-by: Hauke Mehrtens <hauke@hauke-m.de>

olemmela committed:
kernel: modules: add irq support for pca953x

Enable interrupt controller support for PCA953x kernel module

Signed-off-by: Oskari Lemmela <oskari@lemmela.net>
Link: https://github.com/openwrt/openwrt/pull/15926
Signed-off-by: Hauke Mehrtens <hauke@hauke-m.de>

olemmela committed:
kernel: netdevices: add leds support

Compile LED support to swconfig and phy modules

Signed-off-by: Oskari Lemmela <oskari@lemmela.net>
Link: https://github.com/openwrt/openwrt/pull/15926
Signed-off-by: Hauke Mehrtens <hauke@hauke-m.de>

skramstad commented:
I tested ag71xx upstream, but no luck...
```
ag71xx: probe of 19000000.eth failed with error -2
```
Did it work for your device?

john-tho commented:
This is starting to look good, and we finaly have the buildbot capability to build device specific initramfs images.

Have we considered what it would take to make DTS work with both systems (via different `compatible` strings and some duplication if needed). There will always be some who prefer one driver (DTS / switchdev) over the other. OpenWrt network config would be more complicated, but could maybe bridge the switchdev switch-vlan interface and the dsa ports, as long as netdev names don't overlap (no `eth%d` on DSA), and the bridge ignores the missing interfaces.

---

> I tested ag71xx upstream, but no luck...
> ```
> ag71xx: probe of 19000000.eth failed with error -2
> ```

Need to detail your DTS changes if you want any help debugging.

Look for `-ENOENT` in the upstream driver, or anything it calls:
https://elixir.bootlin.com/linux/v6.6.38/source/drivers/net/ethernet/atheros/ag71xx.c#L662
The upstream binding documentation: https://elixir.bootlin.com/linux/v6.6.38/source/Documentation/devicetree/bindings/net/qca,ar71xx.yaml#L27

skramstad commented:
@john-tho sorry about the missing DTS. Added below my current DTS with DSA.
```
&mdio0 {
	status = "okay";

	switch0@0 {
		compatible = "qca,qca8327";
		#address-cells = <1>;
		#size-cells = <0>;

		reg = <0>;

		ports {
			#address-cells = <1>;
			#size-cells = <0>;

			port@0 {
				reg = <0>;
				label = "cpu";
				ethernet = <&eth0>;
				phy-mode = "sgmii";

				fixed-link {
					speed = <1000>;
					full-duplex;
				};
			};

			wan: port@1 {
				reg = <1>;
				label = "wan";
				phy-mode = "internal";
				phy-handle = <&phy_port1>;
				nvmem-cells = <&macaddr_art_0 0>;
				nvmem-cell-names = "mac-address";
			};

			port@2 {
				reg = <2>;
				label = "lan1";
				phy-mode = "internal";
				phy-handle = <&phy_port2>;
				nvmem-cells = <&macaddr_art_0 1>;
				nvmem-cell-names = "mac-address";
			};

			port@3 {
				reg = <3>;
				label = "lan3";
				phy-mode = "internal";
				phy-handle = <&phy_port3>;
				nvmem-cells = <&macaddr_art_0 3>;
				nvmem-cell-names = "mac-address";
			};

			port@4 {
				reg = <4>;
				label = "lan2";
				phy-mode = "internal";
				phy-handle = <&phy_port4>;
				nvmem-cells = <&macaddr_art_0 2>;
				nvmem-cell-names = "mac-address";
			};

			port@5 {
				reg = <5>;
				label = "lan4";
				phy-mode = "internal";
				phy-handle = <&phy_port5>;
				nvmem-cells = <&macaddr_art_0 4>;
				nvmem-cell-names = "mac-address";
			};
		};

		mdio {
			#address-cells = <1>;
			#size-cells = <0>;

			phy_port1: phy@0 {
				reg = <0>;
			};

			phy_port2: phy@1 {
				reg = <1>;
			};

			phy_port3: phy@2 {
				reg = <2>;
			};

			phy_port4: phy@3 {
				reg = <3>;
			};

			phy_port5: phy@4 {
				reg = <4>;

			};
		};
	};
};

&eth0 {
	status = "okay";

	pll-data = <0x03000101 0x00000101 0x00001919>;

	fixed-link {
		speed = <1000>;
		full-duplex;
	};
};

&mdio1 {
	status = "disabled";
};

&eth1 {
	status = "disabled";
};

```

olemmela commented:
> I tested ag71xx upstream, but no luck...
> 
> ```
> ag71xx: probe of 19000000.eth failed with error -2
> ```
> 
> Did it work for your device?
> 

I don't currently have any ath79 device with a serial port connected, so I haven't tested this on any real device yet.
I hope in a couple of weeks I might be able to test this.

At least clocks were missing from ethX devices, which could cause upstream driver ENOENT. Those are now added.

skramstad commented:
I added the missing clocks, but no difference.

olemmela commented:
> I added the missing clocks, but no difference.

@skramstad  Could you try to replace line https://elixir.bootlin.com/linux/v6.6.38/source/drivers/net/ethernet/atheros/ag71xx.c#L641
with this one
`if (ag71xx_is(ag, AR9330) || ag71xx_is(ag, AR9340) || ag71xx_is(ag, QCA9530) || ag71xx_is(ag, QCA9550)) {`

It seems qca9530 and qca9550 are using wrong table.

skramstad commented:
@olemmela Thanks, that did work. I'm on QCA9560 so I added that too.
```C
330 enum ag71xx_type {
331         AR7100,
332         AR7240,
333         AR9130,
334         AR9330,
335         AR9340,
336         QCA9530,
337         QCA9550,
338         QCA9560,
339 };

642         if (ag71xx_is(ag, AR9330) || ag71xx_is(ag, AR9340) || ag71xx_is(ag, QCA9530) || ag71xx_is(ag, QCA9550) || ag71xx_is(ag, QCA9560)) {
```
I also added the missing clocks in my dts too
```
&eth0 {
	status = "okay";
	
	resets = <&rst 9>, <&rst 22>;
	reset-names = "mac", "mdio";
	clocks = <&pll ATH79_CLK_AHB>, <&pll ATH79_CLK_AHB>;
	clock-names = "eth", "mdio";
	
	pll-data = <0x03000101 0x00000101 0x00001919>;

	fixed-link {
		speed = <1000>;
		full-duplex;
	};
};
```

The output
```
[    1.942408] ag71xx 19000000.eth (unnamed net_device) (uninitialized): invalid MAC address, using random address
[    2.631378] ag71xx 19000000.eth eth0: Atheros AG71xx at 0xb9000000, irq 4, mode:mii
[    2.662468] kmodloader: done loading kernel modules from /etc/modules-boot.d/*
[    2.680582] init: - preinit -
[    3.811512] qca8k eth.0:00: configuring for fixed/sgmii link mode
[    3.840323] qca8k eth.0:00: Link is Up - 1Gbps/Full - flow control off
[    3.926719] qca8k eth.0:00 wan (uninitialized): PHY [qca8k-0.0:00] driver [Qualcomm Atheros 8327-B internal PHY] (irq=POLL)
[    4.056726] qca8k eth.0:00 lan1 (uninitialized): PHY [qca8k-0.0:01] driver [Qualcomm Atheros 8327-B internal PHY] (irq=POLL)
[    4.186734] qca8k eth.0:00 lan3 (uninitialized): PHY [qca8k-0.0:02] driver [Qualcomm Atheros 8327-B internal PHY] (irq=POLL)
[    4.316748] qca8k eth.0:00 lan2 (uninitialized): PHY [qca8k-0.0:03] driver [Qualcomm Atheros 8327-B internal PHY] (irq=POLL)
[    4.446727] qca8k eth.0:00 lan4 (uninitialized): PHY [qca8k-0.0:04] driver [Qualcomm Atheros 8327-B internal PHY] (irq=POLL)
[    4.500614] ag71xx 19000000.eth eth0: entered promiscuous mode
[    4.506799] DSA: tree 0 setup
```
I believe some more `QCA9560` code is missing in `ag71xx.c`. 
Now the waiting game is on, to see if the ports stops responding after 5-30 mins or so like the downstream module.

olemmela commented:
> @olemmela Thanks, that did work. I'm on QCA9560 so I added that too.
> 
> 
> I also added the missing clocks in my dts too
> 
> ```
> &eth0 {
> 	status = "okay";
> 	
> 	resets = <&rst 9>, <&rst 22>;
> 	reset-names = "mac", "mdio";
> 	clocks = <&pll ATH79_CLK_AHB>, <&pll ATH79_CLK_AHB>;
> 	clock-names = "eth", "mdio";
> 	
> 	pll-data = <0x03000101 0x00000101 0x00001919>;
> 
> 	fixed-link {
> 		speed = <1000>;
> 		full-duplex;
> 	};
> };
> ```
> 

@skramstad  I think correct clocks should be
``
clocks = <&pll ATH79_CLK_AHB>, <&pll ATH79_CLK_MDIO>;
``

skramstad commented:
> > @olemmela Thanks, that did work. I'm on QCA9560 so I added that too.
> > I also added the missing clocks in my dts too
> > ```
> > &eth0 {
> > 	status = "okay";
> > 	
> > 	resets = <&rst 9>, <&rst 22>;
> > 	reset-names = "mac", "mdio";
> > 	clocks = <&pll ATH79_CLK_AHB>, <&pll ATH79_CLK_AHB>;
> > 	clock-names = "eth", "mdio";
> > 	
> > 	pll-data = <0x03000101 0x00000101 0x00001919>;
> > 
> > 	fixed-link {
> > 		speed = <1000>;
> > 		full-duplex;
> > 	};
> > };
> > ```
> 
> @skramstad I think correct clocks should be `clocks = <&pll ATH79_CLK_AHB>, <&pll ATH79_CLK_MDIO>;`

Thanks. I was not sure if I'm supposed to use
https://github.com/torvalds/linux/blob/master/Documentation/devicetree/bindings/net/qca%2Car71xx.yaml#L82
```
clocks = <&pll 1>, <&pll 2>;
clock-names = "eth", "mdio";
```
vs
i.e (ar9331) https://github.com/torvalds/linux/blob/master/arch/mips/boot/dts/qca/ar9331.dtsi#L126
```
clocks = <&pll ATH79_CLK_AHB>, <&pll ATH79_CLK_AHB>;
clock-names = "eth", "mdio";
```
vs
```
<&pll ATH79_CLK_AHB>, <&pll ATH79_CLK_MDIO>
clock-names = "eth", "mdio";
```
Another thing that made me curious is if I need to add reset-gpio https://github.com/torvalds/linux/blob/master/Documentation/devicetree/bindings/net/dsa/qca8k.yaml#L210
```
reset-gpios = <&gpio 42 GPIO_ACTIVE_LOW>;
```
Because now I'm losing lan1 after around 1h 30min. Before it was from 5 to 30 min ish... (4 out of 5 ports working is better than before).

This is my current changes https://github.com/skramstad/openwrt/commit/1354eedf8f461f42fc77d968c790dff72e2a4c5b

olemmela commented:
> > > @olemmela Thanks, that did work. I'm on QCA9560 so I added that too.
> > > I also added the missing clocks in my dts too
> > > ```
> > > &eth0 {
> > > 	status = "okay";
> > > 	
> > > 	resets = <&rst 9>, <&rst 22>;
> > > 	reset-names = "mac", "mdio";
> > > 	clocks = <&pll ATH79_CLK_AHB>, <&pll ATH79_CLK_AHB>;
> > > 	clock-names = "eth", "mdio";
> > > 	
> > > 	pll-data = <0x03000101 0x00000101 0x00001919>;
> > > 
> > > 	fixed-link {
> > > 		speed = <1000>;
> > > 		full-duplex;
> > > 	};
> > > };
> > > ```
> > 
> > 
> > @skramstad I think correct clocks should be `clocks = <&pll ATH79_CLK_AHB>, <&pll ATH79_CLK_MDIO>;`
> 
> Thanks. I was not sure if I'm supposed to use https://github.com/torvalds/linux/blob/master/Documentation/devicetree/bindings/net/qca%2Car71xx.yaml#L82
> 
> ```
> clocks = <&pll 1>, <&pll 2>;
> clock-names = "eth", "mdio";
> ```
> 
> vs i.e (ar9331) https://github.com/torvalds/linux/blob/master/arch/mips/boot/dts/qca/ar9331.dtsi#L126
> 
> ```
> clocks = <&pll ATH79_CLK_AHB>, <&pll ATH79_CLK_AHB>;
> clock-names = "eth", "mdio";
> ```
> 
> vs
> 
> ```
> <&pll ATH79_CLK_AHB>, <&pll ATH79_CLK_MDIO>
> clock-names = "eth", "mdio";
> ```
> 
> Another thing that made me curious is if I need to add reset-gpio https://github.com/torvalds/linux/blob/master/Documentation/devicetree/bindings/net/dsa/qca8k.yaml#L210
> 
> ```
> reset-gpios = <&gpio 42 GPIO_ACTIVE_LOW>;
> ```
> 
> Because now I'm losing lan1 after around 1h 30min. Before it was from 5 to 30 min ish... (4 out of 5 ports working is better than before).
> 
> This is my current changes [skramstad@1354eed](https://github.com/skramstad/openwrt/commit/1354eedf8f461f42fc77d968c790dff72e2a4c5b)
> 
> Edit: I just tried with
> 
> ```
> clocks = <&pll ATH79_CLK_AHB>, <&pll ATH79_CLK_MDIO>;
> ```
> 
> and lan1 did not respond after 10 minutes. Hmm...

After reading qca95xx datasheets correct clock for MDIO is ATH79_CLK_REF. There is separate pll for ethernet so CLK_AHB might not be correct one either.

It might be good to test downstream ag71xx driver too, which uses pll-data to setup clocks.

reset-gpios is optional. It can be found manually by switching all the gpios one by one and when the switch resets it is the correct gpio pin.

skramstad commented:
OK. Thanks for the input @olemmela. 👍

skramstad commented:
@olemmela I now have an uptime at 11 hours. All ports are still working. The only change I did was to use the external setup instead of internal. I don't know why this works and internal does not.

Clocks are still `clocks = <&pll ATH79_CLK_AHB>, <&pll ATH79_CLK_REF>;`.

olemmela commented:
> @olemmela I now have an uptime at 11 hours. All ports are still working. The only change I did was to use the external setup instead of internal. I don't know why this works and internal does not.
> 
> Clocks are still `clocks = <&pll ATH79_CLK_AHB>, <&pll ATH79_CLK_REF>;`.

By external setup you mean external MDIO configuration?

This makes sense because the most likely problem is some sort of MDIO bus race between ethernet MDIO and normal MDIO. External MDIO bus is locked during ethernet MDIO access (https://github.com/torvalds/linux/commit/526c8ee04bdbd4d8d19a583b1f3b06700229a815)

Changing source clocks just sets different divider for MDIO bus clock rate. ag71xx driver tries to keep MDIO clock under 5Mz. Different MDIO clock rate just changes how often race happens.

After reading documentation ethernet clock is always 100Mhz (qca956x also supports 125Mhz). Source for MDIO clock can be ref, ethernet or 50Mhz (I suppose this is just ethernet/2). The current ath79 clock driver does not support selecting the MDIO source clock. So the value is most likely set in the bootloader.

Correct fix for the clocks would be add support for ethernet and mdio clocks in the ath79 clock driver.

skramstad commented:
Yes, you are correct. I was looking at https://github.com/torvalds/linux/blob/master/Documentation/devicetree/bindings/net/dsa/qca8k.yaml#L129 comparing to https://github.com/torvalds/linux/blob/master/Documentation/devicetree/bindings/net/dsa/qca8k.yaml#L204.

I'm not familiar with clocks and stuff, but as you wrote earlier, it sounds like a solution.

> Correct fix for the clocks would be add support for ethernet and mdio clocks in the ath79 clock driver.

wa-Nadoo commented:
I've found the commit with explanation of downstreaming of ag71xx driver https://github.com/openwrt/openwrt/commit/32726846c85985fbc320d123a7b26b32124cd47a Maybe it can be helpful.

neheb commented on target/linux/ath79/dts/qca9558_tplink_archer-c.dtsi:
```diff
@@ -115,36 +116,189 @@
+				reg = <0>;
+				label = "cpu";
+				ethernet = <&eth0>;
+				phy-mode = "rgmii-id";
```
original initvals specify port0 sgmii, port 6 rgmii-id.

olemmela replied:
Yep you are correct. I'm testing different configs if they make any difference. 0x07600000 actually means RGMII with 1ns tx and 2ns rx delays, but those are also defaults in qca8k if delays are not set.

It may be that only port 0 is used for Ethernet MDIO. When you enable only the RGMII port connected to port 6, I don't see any extra Ethernet frames. Also all ports work without problems. I need to debug this a bit more when MDIO over Ethernet is actually enabled.

neheb replied:
According to @Ansuel, using external MDIO or not using SGMII disables mgmt eth. This is a work around. It does not fix the fundamental issue. It should also be slower.

hauke commented on target/linux/ath79/files/drivers/net/ethernet/atheros/ag71xx/Makefile:
```diff
@@ -2,12 +2,12 @@
-obj-$(CONFIG_AG71XX)	+= ag71xx_mdio.o
-obj-$(CONFIG_AG71XX)	+= ag71xx.o
+obj-$(CONFIG_AG71XX_LEGACY)	+= ag71xx_ds_mdio.o
+obj-$(CONFIG_AG71XX_LEGACY)	+= ag71xx_ds.o
```
For some parts you added `legacy` and for others you added `ds`. I would prefer the same extension everywhere. I would even prefer `legacy` if it does not get too long.

olemmela replied:
renamed module name to legacy

hauke commented on target/linux/ath79/files/drivers/net/ethernet/atheros/ag71xx/Makefile:
```diff
@@ -2,12 +2,12 @@
 
-obj-$(CONFIG_AG71XX)	+= ag71xx_mdio.o
-obj-$(CONFIG_AG71XX)	+= ag71xx.o
+obj-$(CONFIG_AG71XX_LEGACY)	+= ag71xx_ds_mdio.o
```
I think we should merge `ag71xx_ds_mdio.ko`  into `ag71xx_ds.ko` I do not see any reson why this should be a seperate module.

olemmela replied:
Is there any good reason to change the code of the legacy driver, or should all the effort be put into implementing the missing parts in the upstream driver?

hauke replied:
An extra kernel module needs some 100 bytes more space compared to compiling it into one. We always use both modules together.

hauke commented:
How much is the size of the root file system increasing after the code is compiled as a module?

neheb commented on target/linux/ath79/dts/qca955x.dtsi:
```diff
@@ -379,4 +381,6 @@
 
 	resets = <&rst 13>, <&rst 23>;
 	reset-names = "mac", "mdio";
+	clocks = <&pll ATH79_CLK_AHB>, <&pll ATH79_CLK_REF>;
```
this is interesting.

Is this compatible with the downstream driver as well?

https://github.com/openwrt/openwrt/commit/32726846c85985fbc320d123a7b26b32124cd47a says something about the upstream driver not supporting clocks. I assume this means these sections would be ignored with the downstream driver.

I ask as it would be interesting to switch to the upstream driver in one PR and DSA module conversion in another. OTOH that might not really make sense. I don't even know if the upstream driver works with swconfig.

olemmela replied:
The dowstream driver sets the clocks directly, so it does not use the values ​​of DTS.

The upstream clock driver does not support Ethernet or MDIO clocks at all. The current upstream ag71xx driver only enables the ethernet clock and does not configure it. The MDIO clock is only used to calculate the clock divider of the MDIO bus clock.

The Ethernet clock can be configured to 100Mhz or 125Mhz and the MDIO clock can be REF, 50Mhz or 100Mhz. These bits are set in the bootloader, so there is no need to change them on the linux side. This is why setting the ethernet clock to only something that is already in use and the MDIO clock to REF works without proper clock support.

I haven't tested if upstream driver works with swconfig

olemmela commented:
> How much is the size of the root file system increasing after the code is compiled as a module?

rootfs compressed size:
(3748kb) swconfig without modules
(3828kb) swconfig with modules (+80kb)
(3858kb) DSA with modules (+110kb)

hauke commented on target/linux/ath79/config-6.6:
```diff
@@ -169,15 +157,12 @@ CONFIG_PCI_DOMAINS=y
 CONFIG_PERF_USE_VMALLOC=y
 CONFIG_PGTABLE_LEVELS=2
-CONFIG_PHYLIB=y
-CONFIG_PHYLIB_LEDS=y
```
When is the phylib not needed on ath79? 
If we need an option in all scenarios on a SoC we should compile it into the kernel becasue that saves some space compared to building it as a module.

olemmela replied:
Added phylib back

hauke commented on package/kernel/linux/modules/netdevices.mk:
```diff
@@ -407,13 +524,27 @@ endef
 define KernelPackage/dsa-tag-dsa
   SUBMENU:=$(NETWORK_DEVICES_MENU)
   TITLE:=Marvell DSA type DSA and EDSA taggers
+  DEPENDS:=+kmod-dsa +kmod-regmap-core
```
Why does this depend on `kmod-regmap-core` now?

olemmela replied:
This was maybe rebase mistake. It is now removed.

hauke commented on package/kernel/linux/modules/netdevices.mk:
```diff
@@ -343,6 +427,23 @@ endef
+   SUBMENU:=$(NETWORK_DEVICES_MENU)
+   TITLE:=Micrel PHYs
+   KCONFIG:=CONFIG_MICREL_PHY
+   DEPENDS:=+kmod-libphy +kmod-ptp
```
Do we always need the dependency to `kmod-ptp`? The target only set `CONFIG_PTP_1588_CLOCK_OPTIONAL` in its configuration, this dependency will add the full ptp framework. 

I do not know how to fix this. I think we had similar problems already in other places.

hauke replied:
maybe we can also ignore this for now and hope that this kernel package is only added to devices which are actually using the micrel phy chip.

hauke commented on target/linux/ixp4xx/config-6.6:
```diff
@@ -85,6 +85,7 @@ CONFIG_FORCE_PCI=y
 CONFIG_FS_IOMAP=y
 CONFIG_FS_MBCACHE=y
 CONFIG_FUNCTION_ALIGNMENT=0
+# CONFIG_FSL_DPAA2_SWITCH is not set
```
Why is this change needed? Why is it only needed for these two targets?

olemmela replied:
Maybe this was related to older kernel versions or something else have changed in master. But anyway seems to work without it.

hauke commented:
> > How much is the size of the root file system increasing after the code is compiled as a module?
> 
> rootfs compressed size: (3748kb) 
> swconfig without modules (3828kb)
> swconfig with modules (+80kb)
> (3858kb) DSA with modules (+110kb)

The size increases are pretty big also when you compare it to a size reduction of 94 KB for the uncompressed(?) kernel. 

When we use kernel modules we can remove some PHY drivers from device images where they are not needed, we can gain some space like this. 

I assume with that optimization the images will probably be smaller. Some devices also have a constraint for the kernel image size, that is also improved with this change. 

I would merge this after the minor comments are solved with converting the device to DSA, that can be done in a separate PR. 
Could you split the last commit into two, one doing the general improvements and one adding the DTS and image changes to the board.

olemmela committed:
ath79: qca955x ag71xx upstream driver fix

Fix mdio probe in qca95xx devices and add mandatory clocks to dtsi

Signed-off-by: Oskari Lemmela <oskari@lemmela.net>
Link: https://github.com/openwrt/openwrt/pull/15926
Signed-off-by: Hauke Mehrtens <hauke@hauke-m.de>

olemmela commented:
> > > How much is the size of the root file system increasing after the code is compiled as a module?
> > 
> > 
> > rootfs compressed size: (3748kb)
> > swconfig without modules (3828kb)
> > swconfig with modules (+80kb)
> > (3858kb) DSA with modules (+110kb)
> 
> The size increases are pretty big also when you compare it to a size reduction of 94 KB for the uncompressed(?) kernel.
> 
> When we use kernel modules we can remove some PHY drivers from device images where they are not needed, we can gain some space like this.
> 
> I assume with that optimization the images will probably be smaller. Some devices also have a constraint for the kernel image size, that is also improved with this change.
> 
> I would merge this after the minor comments are solved with converting the device to DSA, that can be done in a separate PR. Could you split the last commit into two, one doing the general improvements and one adding the DTS and image changes to the board.

I splitted last commit and kept only ag71xx driver fixes. Device DSA migrations can be done in separate PRs.

There are still a couple of issues with DSA and the ath79 target.
 - qca8k ethernet mdio not working properly. A couple of different workarounds are possible to disable the feature, but not an actual fix.
 - Ethernet controller does not support checksumming, so GRO is terribly slow. The workaround is to disable GRO

neheb commented:
Disabling GRO also applies to swconfig I believe.

neheb commented:
@oskarirauta got a commit for that archer c7 v2 conversion?

oskarirauta commented:
I don’t have archer.. Person mix up?Lähetetty iPhonestaRosen Penev ***@***.***> kirjoitti 9.8.2024 kello 2.14:﻿
@oskarirauta got a commit for that archer c7 v2 conversion?

—Reply to this email directly, view it on GitHub, or unsubscribe.You are receiving this because you were mentioned.Message ID: ***@***.***>

olemmela commented:
> @oskarirauta got a commit for that archer c7 v2 conversion?

https://github.com/olemmela/openwrt/commit/5e6e5cb1be8220808215dfc0aeff091a95041aa5

SGMII port is commented out. Just remove comments if you want to test MDIO over Ethernet

olemmela commented:
> Disabling GRO also applies to swconfig I believe.

I think ag71xx driver does not currently do GRO because of old NAPI or something. DSA uses gro cells, so that is difference why swconfig is not affected.

neheb commented:
> > Disabling GRO also applies to swconfig I believe.
> 
> I think ag71xx driver does not currently do GRO because of old NAPI or something. DSA uses gro cells, so that is difference why swconfig is not affected.

GRO needs hardware checksumming for speed. The archer has an external NAT chip that can do this, but nobody has written the code, mostly because the HW offload framework was not available.

olemmela commented:
@hauke is there still something which blocks this to be merged?

I don't see any real benefit rewriting legacy driver MDIO probe just for saving 100bytes.

hauke commented:
I took the patches which add the modules, but not the one which switches the ath79 target: https://github.com/openwrt/openwrt/pull/15926/commits/111524f6ae37058ff56b81a6b9fbe115d5ac7fee 

This last patch looks a bit dangerous to me.

rsalvaterra commented:
I just looked at bdd2d685db3620298928b1b0b2a565475ce16b2e, since I was scratching my head about `kmod-dsa`'s dependencies. Why is it selecting `kmod-mdio-devres` and `kmod-net-selftests` when, according to the kernel configuration, DSA support doesn't require them at all?

luizluca commented:
@olemmela, what is the difference between ag71xx-legacy (downstream) and the ag71xx (upstream). I tried to dig that info from the commit messages, description and this thread but I couldn't find it (maybe it is in a code comment). Should we use the former for swconfig and the latter for DSA?

Also, I wonder why ag71xx-legacy ended up in package/kernel/linux/modules/netdevices.mk and not in target/linux/ath79/modules.mk .

neheb commented:
It's upstream vs downstream. Upstream lacks some functionality.
